src/mapogcsld.cpp

```diff
@@ -269,202 +525,30 @@ int msSLDApplySLD(mapObj *map, const char *psSLDXML, int iLayer,
             goto sld_cleanup;
           }
 #endif
-
-          if (sldLayer->numclasses > 0) {
-            int iClass = 0;
-            int k;
-            int bSLDHasNamedClass = MS_FALSE;
-
-            lp->type = sldLayer->type;
-            lp->rendermode = MS_ALL_MATCHING_CLASSES;
-
-            for (k = 0; k < sldLayer->numclasses; k++) {
-              if (sldLayer->_class[k]->group) {
-                bSLDHasNamedClass = MS_TRUE;
-                break;
-              }
-            }
-
-            for (k = 0; k < lp->numclasses; k++) {
-              if (lp->_class[k] != NULL) {
-                lp->_class[k]->layer = NULL;
-                if (freeClass(lp->_class[k]) == MS_SUCCESS) {
-                  msFree(lp->_class[k]);
-                  lp->_class[k] = NULL;
-                }
-              }
-            }
-            lp->numclasses = 0;
-
-            if (bSLDHasNamedClass && sldLayer->classgroup) {
-              /* Set the class group to the class that has UserStyle.IsDefault
-               */
-              msFree(lp->classgroup);
-              lp->classgroup = msStrdup(sldLayer->classgroup);
-            } else {
-              /*unset the classgroup on the layer if it was set. This allows the
-              layer to render with all the classes defined in the SLD*/
-              msFree(lp->classgroup);
-              lp->classgroup = NULL;
-            }
-
-            for (k = 0; k < sldLayer->numclasses; k++) {
-              if (msGrowLayerClasses(lp) == NULL)
+          if (msApplySldLayerToMapLayer(sldLayer, lp) == MS_FAILURE) {
+            return MS_FAILURE;
+          };
+          bSldApplied = 1;
+          break;
+        }
+      }
+      if (bUseSpecificLayer) {
+        if (bSldApplied == 0) {
+          // there was no name match between the map layer and the SLD named
+          // layer - check if we should apply the first SLD layer anyway
+          const char *pszSLDUseFirstNamedLayer =
+              msLookupHashTable(&(lp->metadata), "SLD_USE_FIRST_NAMEDLAYER");
+          if (pszSLDUseFirstNamedLayer) {
+            if (strcasecmp(pszSLDUseFirstNamedLayer, "true") == 0) {
+              sldLayer = &pasSLDLayers[0];
```

Comment:
```suggestion
              layerObj *sldLayer = &pasSLDLayers[0];
```

Reply:
Is it ok to re-declare the same variable name in a nested block (it is also declared on line 498), or should a different variable name be used?

Reply:
> Is it ok to re-declare the same variable name in a nested block (it is also declared on line 498)

sldLayer = &pasSLDLayers[j]; at line 501 seems to be in a unrelated scope of the one at line 543. Or perhaps I don't read correctly the scopes in the github review UI?

Reply:
Yes different scope (assuming C doesn't use hoisting like JS). I've renamed the variable anyway for clarity.